torch/utils/data/datapipes/iter/combining.py

```diff
@@ -0,0 +1,183 @@
+from collections import defaultdict, deque
+from torch.utils.data import IterDataPipe
+from typing import Iterator, Optional, Sequence, Sized, Tuple, TypeVar
+
+T_co = TypeVar('T_co', covariant=True)
+
+
+class UnzipIterDataPipe(IterDataPipe):
+    r""" :class:`UnzipIterDataPipe`.
+
+    The inverse of :class:`ZipIterDataPipe`, this class disaggregates
+    the elements of the zipped iterable DataPipe. The first element
+    determines the number of output DataPipe. A :class:`ZipIterDataPipe`
+    is often attached later to aggregate these output DataPipes back.
+    Return a tuple of split DataPipes.
+
+    Example:
+        >>> print(list(dp))
+        [(1, 2), (3, 4), (5, 6), ...]
+        >>> dp1, dp2 = datapipes.iter.Unzip(dp)
+        >>> print(list(dp1))
+        [1, 3, 5, ...]
+        >>> print(list(dp2))
+        [2, 3, 4, ...]
+        >>> def fn(data):
+        ...     return torch.tensor(data)
+        >>> map_dp = datapipes.iter.Map(dp1, fn=fn)
+        >>> zipped_dp = datapipes.iter.Zip(map_dp, dp2)
+        >>> print(list(zipped_dp))
+        [(tensor(1), 2), (tensor(3), 4), (tensor(5), 6), ...]
+
+    .. note:: This DataPipe may require significant storage, especially
+              when one iterator uses most of the data before another
+              iterator starts.
+
+    args:
+        datapipe: Iterable DataPipe being disaggregated
+    """
+    datapipe: IterDataPipe[Sequence]
+    num_splits: int
+    _it: Iterator[Sequence]
+
+    def __new__(cls, datapipe: IterDataPipe[Sequence]):
+        source_dp = super().__new__(cls)
+        it = iter(datapipe)
+        try:
+            data = next(it)
```

Comment:
It is not the best idea to start iterating items before completion of graph. I assume you want to type check on construction, but please move this check to runtime.

Reply:
You are right, it makes sense not to start iteration at the construction time.
But, it raises another problem for type checking. I am wondering do you have any suggestion for all type checking at the construction time, if we do want this feature, we may need to force each DataPipe to specify it's own data type. Otherwise, most of type checking needs to happen at the runtime.

Reply:
However, for `Unzip`, we need to know the number of elements to split the input DataPipe into pieces in the construction time. So, we have to fetch one data from input. But, we wont't need to run this duplicate iteration if I pass the first data and iterator to __init__.
I will update the PR and let you know.